### PR TITLE
Versionless apps validation and dedupe listeners for app events

### PR DESCRIPTION
**Pull request by diegoribero**
- Validate if an app has version details before trying to download it (offline sync, failed for vehicledetector custom app)
- Instead of adding multiple listeners each time an app event is registered, add one listener with a collection of callbacks per event

### CODE REVIEW

File: apps/matrix.js [hunk 1]
```diff
@@ -152,25 +154,35 @@ function interAppResponse(name, cb) {
...
-  console.log('setup event listeners:', name);
+  
+  if (_.isString(name)) {
+    console.log('setup event listeners:', name);
+    eventCbs[name] = cb; //Store callback for event
+  } else {
+    console.log('setup nameless event listener');
+  }
+
+  if (!eventCheck) { //This makes sure to only add one message listener that handles all messages
```
Comment (eighteyes) [inline apps/matrix.js]: well done 👏

File: apps/matrix.js [hunk 2]
```diff
@@ -23,6 +23,8 @@ error = function() {
 
 var appName = '';
 var assetPath = '';
+var eventCheck = false;
+var eventCbs = {};
```
Comment (eighteyes) [inline apps/matrix.js]: An explanation of the data structure here would be helpful.

File: apps/matrix.js [hunk 3]
```diff
@@ -152,25 +154,35 @@ function interAppResponse(name, cb) {
...
-    // is global or app-specific
-    if (m.type === 'trigger' || m.type === 'app-message' || m.type === 'app-' + appName + '-message') {
-      console.log('[M]->app(msg)'.blue, m);
-      if (_.isString(name)) {
         // if an event name was specified in the on()
-        if (m.eventName === name || m.value === name ) {
-          cb(m);
+        if (eventCbs[m.eventName]) {
+          eventCbs[m.eventName](m);
+        } else if (eventCbs[m.value]) {
```
Comment (eighteyes) [inline apps/matrix.js]: some comments here about routing decisions would be helpful

**Comment by diegoribero** [comment]
Added some more inline comments to explain the process

**Pull request merged**